pkg/cluster/context.go

```diff
@@ -0,0 +1,20 @@
+package cluster
+
+import "context"
+
+// contextKey is how we find Loggers in a context.Context.
+type contextKey struct{}
+
+// FromContext returns a Logger from ctx or an error if no Logger is found.
+func FromContext(ctx context.Context) string {
+	if v, ok := ctx.Value(contextKey{}).(string); ok {
+		return v
+	}
+	return ""
+}
+
+// IntoContext returns a new Context, derived from ctx, which carries the
+// provided cluster identifier.
+func IntoContext(ctx context.Context, cluster Cluster) context.Context {
+	return context.WithValue(ctx, contextKey{}, cluster.ID())
```

Comment:
This implicitly assumes that a controller will always interact with exactly one cluster, which often isn't true

Reply:
The idea behind the context key is to have the request cluster information always available, Clients would know which client to use based on this information.

We can, however, decide to go to a complete different path and rework our clients to be cluster-scoped, and in turn have the request coming into a Reconcile loop to be enhanced with other means to get clients/mappers/rest/etc